sokoban.hs

```diff
@@ -198,55 +202,68 @@ play (QuitGame s) =
   do
     putStrLn "Exiting level..."
 
+play (UndoGame (State board pS (r, c) l)) =
+  do
+    putStrLn "Switching to previous move..."
+    if pS == Empty then (putStrLn "No previos move...")
+    else play (ContinueGame pS)
+     
+play (RestartGame (State board pS (r, c) l))=
+  do
+    putStrLn "Restarting level..."
+    playLevel l
+
 -- takes user input and current state of board
 -- returns the next state of board as a result (game won or continue game)
 -- should be used as a helper in the main game function
-getNextBoard :: Game
-getNextBoard move (State board (r,c))
-  | move == (Action 'W') = movePlayer (State board (r,c)) (r-1,c) (r-2,c)
-  | move == (Action 'A') =  movePlayer (State board (r,c)) (r,c-1) (r,c-2)
-  | move == (Action 'S') = movePlayer (State board (r,c)) (r+1,c) (r+2,c)
-  | move == (Action 'D') = movePlayer (State board (r,c)) (r,c+1) (r,c+2)
-  | move == (Action 'Q') = QuitGame (State board (r, c))
-  | otherwise = ContinueGame (State board (r, c)) -- return same state
+getNextBoard :: Game 
+getNextBoard move (State board pS (r,c) l)
+  | move == (Action 'W') = movePlayer (State board (State board pS (r,c) l) (r,c) l) (r-1,c) (r-2,c)
+  | move == (Action 'A') = movePlayer (State board (State board pS (r,c) l) (r,c) l) (r,c-1) (r,c-2)
+  | move == (Action 'S') = movePlayer (State board (State board pS (r,c) l) (r,c) l) (r+1,c) (r+2,c)
+  | move == (Action 'D') = movePlayer (State board (State board pS (r,c) l) (r,c) l) (r,c+1) (r,c+2)
+  | move == (Action 'Q') = QuitGame (State board (State board pS (r,c) l) (r, c) l)
+  | move == (Action 'R') = RestartGame (State board (State board pS (r,c) l) (r, c) l)
+  | move == (Action 'U') = UndoGame (State board pS (r, c) l)
+  | otherwise = ContinueGame (State board (State board pS (r,c) l) (r, c) l) -- return same state
```

Comment:
Might be cleaner if we define 
`where newPrevState = (State board pS (r,c) l)`

Reply:
ok